### DISCUSSION TITLE
Unknown issue in using MotionManager interface

### DISCUSSION THREAD

**Discussion by PeterCalifano** [Yaskawa-Global/YMConnect#10] [Q&A] [answered]
Hello,
this will be a rather long message, I apologize in advance.
Me and my colleagues have been using YMConnect for some weeks with a YRC1000 controller and found that it is really effective in easing the use of the robots for our application. A necessary premise is that we're definitely not experts of the field. Unfortunately, an issue is currently preventing us to continue the development. Let me give some details.

We have two Yaskawa GP12 robots controlled by a YRC1000. The robots are twins, with the coordinated motion option function enabled. I know this is not supported yet by YMConnect, but we currently don't need it (i.e., controlling each robot separately is ok for us). I have configured the controller according to the documentation at https://developer.motoman.com/en/YMConnect/RobotControllerSetup. Namely, the controller parameters are set:

RS0 = 2 
RS5 = 1
RS7 = 2 
RS22 = 1 
RS29 = 1 
RS34 = 200 
RS35 = 200
S2C541 = 0 
S2C542 = 0 

All the library classes I have tested so far are working (mainly connection, reading data/info and executing a pre-programmed job from the job list). The only exception is MotionStart() in Motion Manager Interface. I have tested it both specifying the target position (PositionData) manually and copying it from the output of ReadPositionData. I include a snapshot of the code of the latter case for clarity.

```C++

  PositionData robotPositionData{}; 
  int userFrameID = 0;
  int toolNum = 0;
  CoordinateType coordType = CoordinateType::BaseCoordinate;

  readStatus = c->ControlGroup->ReadPositionData(ControlGroupId::R1, coordType, userFrameID, toolNum, robotPositionData);

  PositionData destinationData = robotPositionData; // Copy content of current position into destination
  destinationData.axisData[0] -= 20.0; //Add random small displacement
  destinationData.axisData[1] += 10.0; 

  status = c->ControlCommands->SetServos(SignalStatus::ON);
  cout << "SERVOS ACTIVATION status: " << status << endl;

  JointMotion r1Motion(ControlGroupId::R1, destinationData, 10.0, MotionAccelDecel());
  status = c->MotionManager->AddPointToTrajectory(r1Motion);
  cout << "\nTrajectory point added: " << status << endl;

  // TEST MOTION
  status = c->MotionManager->MotionStart();
  cout << "\nMOTION EXECUTION status: " << status << endl;
  system("sleep 3");
  status = c->MotionManager->MotionStop();
```

Everything goes smoothly till the call of MotionStart(), and the strange thing is that it actually returns an OK status but the robot never moves. What happens is that an alarm is triggered with code 8003[223] whose description is really "there is no description of this error".
In fact, there is no information about it at all (neither in Yaskawa manuals nor the web). The only clue I have found is in the motoROS documentation (at http://wiki.ros.org/motoman_driver/Troubleshooting#Alarm:_8003.5B11.5D), leading us to think this is a not yet addressed configuration error, since [223] is not present among the variants of the 8003 alarm. 
Therefore, we deemed necessary to ask for support as we have no clue about what's going on. The only idea I'm not able to verify is that some parameters of the coordinated motion function is interfering with the proper functioning of MotionManager in a weird way. Thanks for any help.

Pietro Califano,
PhD student, Politecnico di Milano, DART Lab

**Comment by ted-miller** [comment]
Hello @PeterCalifano 

The code snippet you posted looks OK. 

The alarm code `8003` can only be generated from either an INFORM job or a MotoPlus application. So, I suspect that there is another MotoPlus application installed which is interfering with motion. 

I'm not familiar with any applications that use the subcode `[223]`. Is there any additional text when the alarm code appears?

Can you post a copy of the `PANELBOX.LOG`, `ALL.PRM`, and `ALMHIST.DAT` from your robot controller? The PANELBOX should contain a listing of any MotoPlus applications that are currently running. The PRM file is to verify configuration. And the DAT file is just to see the alarm history.

**Reply by PeterCalifano** [reply]
Hello @ted-miller

Really thanks for the quick response.
I have retrieved the files you mentioned before doing the test I describe below. They are attached to this message.

Regarding the alarm there is really no description. The controller just outputs something like "no description found for this error", which is also why we were a bit confused.

Then, I may have a little interesting update about your suspect. I have disabled the option function "MotoPlus FUNC" through the maintenance mode. This still does not solve the problem, but it seems that no error `8003` shows up anymore on the controller side when MotionStart() is called. What happens now is that the method returns an error status with code `65535` and message `General Error. Verify Input parameters`.

[ALL.PRM.txt](https://github.com/Yaskawa-Global/YMConnect/files/14905335/ALL.PRM.txt)
[ALMHIST.DAT.txt](https://github.com/Yaskawa-Global/YMConnect/files/14905337/ALMHIST.DAT.txt)
[PANELBOX.LOG.txt](https://github.com/Yaskawa-Global/YMConnect/files/14905338/PANELBOX.LOG.txt)

**Comment by aseamans** [comment]
Hello @PeterCalifano 
When you get a chance can you upload your controller CMOS so I can test this out? 
Thank you.
-Austen

**Reply by ted-miller** [reply, marked as answer]
@PeterCalifano 
We believe the issue is base-track that's mounted to R1. Since you are requesting a motion in `base-frame`, you must also specify a target for the base axis. 

When calling `AddPointToTrajectory`, there are optional arguments for up to four control groups. Please provide a second argument that has a target position for the linear rail. 

Alternatively, for a quick test, you could try targeting R2 instead of R1. (Or target `robot-frame` for R1.)

If you could also provide the CMOS.BIN file requested by aseamans, that would still be helpful. We would like to reproduce your exact scenario and improve the generic error message that you received, `65535 General Error. Verify Input parameters`.
It will be improved to be more specific about what parameters/arguments are missing.

-----------------------------

Additionally, we're not 100% certain about the behavior of `system("sleep 3");`. It might work just fine, but we want to ensure that only the current thread is sleeping and not the whole programs. If the whole program sleeps (including background threads), then the motion targets will not be transmitted to the robot controller. 

We recommend something like: `std::this_thread::sleep_for(std::chrono::seconds(3));`

**Reply by PeterCalifano** [reply]
Hello again,

I completely forgot to mention the rail before, sorry. In addition, another detail that could be useful is that alarm `4754` occurs frequently when the controller is ON when using the parameters configuration requested by the library.

Concerning the tests you suggested, we already tested using `robot-frame` for R1 only and the same for R2 (both robot and base) before solving the alarm `8003` today. I tested again as you suggested after having completely removed the "sleep" command, but no result yet due to the same error. 

I'm going to try adding the rail as second control group in `AddPointToTrajectory` and get back to you with the result. This something we previously failed to recognize, because we considered it as an extension of R1 group (like a 7th axis). However, it is not so in practice when using YMConnect. In the meanwhile, I already managed to read ControlGroup::B1, which in fact should be the base-track. 

Here is our CMOS.bin as zip file:
[CMOS.zip](https://github.com/Yaskawa-Global/YMConnect/files/14908082/CMOS.zip)

**Reply by ted-miller** [reply]
Hi @PeterCalifano,

@aseamans and I are still working on solving this. We have a copy of your CMOS loaded into a debug controller. There are couple issues.

The `4754`, `4758`, and `4752` alarms are related to the Machine Safety board in your robot controller. Unfortunately, we will not be able to solve these alarms in this forum. All of the diagnostic steps point toward a failure with the ASF01 board. Please contact your local Yaskawa support office for assistance with those alarms.

Knowing that we would not be able to solve those, I disabled all of the Machine Safety functions in your configuration. Unfortunately, we are still unable to achieve motion with this configuration. We're working to identify and isolate which particular setting(s) would prevent the motion API from executing.

In order to help us understand what may be unique with your robot's configuration, can you tell me what MotoPlus application(s) was installed previously?

(If the `MotoPlus Func` is enabled, then you can view these in maintenance mode under `MotoPlus APL` > `File List`. Alternatively, you can just look at the contents of the SD card that's inserted in the ACP01 board of the robot controller. The files will be contained in the `APPLICATION` folder.)

**Reply by ted-miller** [reply]
Ah, I think I just found the problem. It appears that `MotoPlus Func` must be enabled for the motion API's to work. Please re-enable this function. 

I understand that you'll probably then get the `8003[223]` alarm again when you enable the function. You'll need to remove the MotoPlus application that is generating that alarm.

**Reply by PeterCalifano** [reply]
Hello @ted-miller ,

Some great news on our side. In short, we managed to make it work. I'm going to give you the details next, it may be useful.

Firstly, let me clarify that the three alarms do not appear all the time. We are currently unaware of the specific cause triggering them, but we're going to investigate it. Nevertheless, everything seems to work fine.
I enabled the `MotoPlus Func` again as you suggested and checked the applications running with MotoPlus. However, there is none. The problem was not really there, but in something you mentioned before, namely the rail.
The error 8003[223] only shows up when we try to command either of the two robots, regardless of the coordinates type, _**without**_ specifying a motion command for the rail as second target in AddPointToTrajectory(). This is why the test we made some days ago failed, even though we thought it should have not.

I think we can mark this issue as closed. Let me know if any additional detail can be useful for development purposes. Thanks a lot.

Pietro

**Reply by aseamans** [reply]
when this alarm occurs, is it after a successful call to MotionStart()?

**Reply by ted-miller** [reply]
> In short, we managed to make it work

Great news.

> The error 8003[223] only shows up when we try to command either of the two robots

aseamans has located the source of the alarm. We're going to fix it up so that it gives a proper message. Additionally, we will be documenting the error codes for the future.

**Comment by aseamans** [comment]
@PeterCalifano Just a quick update.
I've loaded up your CMOS and am getting the same issue.
I'm working on some troubleshooting on my end and will update you as soon as I have some new information.
-Austen
